File: master_plan_it/master_plan_it/report/mpit_year_comparison/mpit_year_comparison.js
frappe.query_reports["MPIT Year Comparison"] = {
    filters: [
        { fieldname: "year_a", label: __("Year A"), fieldtype: "Link", options: "MPIT Year", default: String(new Date().getFullYear() - 1), reqd: 1 },
        { fieldname: "year_b", label: __("Year B"), fieldtype: "Link", options: "MPIT Year", default: String(new Date().getFullYear()), reqd: 1 },
        { fieldname: "group_by", label: __("Group By"), fieldtype: "Select", options: "Cost Center\nVendor\nProject\nContract\nFunding\nProject Stage", default: "Cost Center" },
        { fieldname: "basis", label: __("Basis"), fieldtype: "Select", options: "Year-end Forecast", default: "Year-end Forecast" },
        { fieldname: "only_changed", label: __("Only Changed"), fieldtype: "Check", default: 1 },
        { fieldname: "min_delta_amount", label: __("Min Delta Amount"), fieldtype: "Currency", default: 0 },
        { fieldname: "min_delta_percent", label: __("Min Delta %"), fieldtype: "Percent", default: 5 },
        { fieldname: "show_new_items", label: __("Show New Items"), fieldtype: "Check", default: 1 },
        { fieldname: "show_removed_items", label: __("Show Removed Items"), fieldtype: "Check", default: 1 },
        { fieldname: "print_profile", label: __("Print Profile"), fieldtype: "Select", options: "Executive\nStandard\nDetailed\nAudit", default: "Standard" },
        { fieldname: "print_orientation", label: __("Print Orientation"), fieldtype: "Select", options: "Auto\nPortrait\nLandscape", default: "Auto" },
        { fieldname: "print_density", label: __("Print Density"), fieldtype: "Select", options: "Normal\nCompact", default: "Normal" },
    ],

    formatter(value, row, column, data, default_formatter) {
        const formatted = default_formatter(value, row, column, data);
        if (column.fieldname === "status" && value === "Critical") {
            return `<span class="text-danger fw-semibold">${formatted}</span>`;
        }
        if (column.fieldname === "status" && value === "Warning") {
            return `<span class="text-warning fw-semibold">${formatted}</span>`;
        }
        if (column.fieldname === "recommended_action" && ["Expired", "Renegotiate"].includes(value)) {
            return `<span class="text-danger fw-semibold">${formatted}</span>`;
        }
        return formatted;
    },
};

File: master_plan_it/master_plan_it/report/mpit_year_comparison/mpit_year_comparison.py
from __future__ import annotations

import frappe
from frappe import _

from master_plan_it.master_plan_it.financial_engine import get_year_comparison_dataset


def execute(filters=None):
    filters = frappe._dict(filters or {})
    dataset = get_year_comparison_dataset(filters)
    return _get_columns(filters), dataset.get("rows", []), None, dataset.get("chart"), _get_report_summary(dataset.get("summary", {}))


def _get_columns(filters) -> list[dict]:
    return [
        {"label": _("Group"), "fieldname": "group_label", "fieldtype": "Data", "width": 220},
        {"label": _("Year A"), "fieldname": "amount_a", "fieldtype": "Currency", "width": 130},
        {"label": _("Year B"), "fieldname": "amount_b", "fieldtype": "Currency", "width": 130},
        {"label": _("Delta Amount"), "fieldname": "delta_amount", "fieldtype": "Currency", "width": 140},
        {"label": _("Delta %"), "fieldname": "delta_percent", "fieldtype": "Percent", "width": 110},
        {"label": _("Change Type"), "fieldname": "change_type", "fieldtype": "Data", "width": 120},
        {"label": _("Driver"), "fieldname": "main_driver", "fieldtype": "Data", "width": 130},
    ]


def _get_report_summary(summary: dict) -> list[dict]:
    return [
        {"label": _("Year A"), "value": summary.get("amount_a", 0), "datatype": "Currency", "indicator": "Blue"},
        {"label": _("Year B"), "value": summary.get("amount_b", 0), "datatype": "Currency", "indicator": "Orange"},
        {"label": _("Delta"), "value": summary.get("delta_amount", 0), "datatype": "Currency", "indicator": "Red"},
        {"label": _("Changed Rows"), "value": summary.get("changed_rows", 0), "datatype": "Int", "indicator": "Grey"},
    ]
